Make sure that frozenlists are pickle-able

csm10495 opened the pull request:
<!-- Thank you for your contribution! -->

## What do these changes do?

Fixes bug where frozenlists can't be pickled.

## Are there changes in behavior for the user?

Not directly. Though this fixes a behavior change introduced after 1.6.0

## Related issue number

#683

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] If you provide code modifications, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep the list in alphabetical order, the file is sorted by name.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<category>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure category is one of the following:
    * `.bugfix`: Signifying a bug fix.
    * `.feature`: Signifying a new feature.
    * `.breaking`: Signifying a breaking change or removal of something public.
    * `.doc`: Signifying a documentation improvement.
    * `.packaging`: Signifying a packaging or tooling change that may be relevant to downstreams.
    * `.contrib`: Signifying an improvement to the contributor/development experience.
    * `.misc`: Anything that does not fit the above; usually, something not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

csm10495 committed:
Make sure that frozenlists are pickle-able

Add contributors / changes updates

Add missing file

csm10495 commented on CHANGES/650.contrib:
```diff
@@ -1,3 +1,2 @@
-The pull request template to be consistent with :ref:`the contributing guidelines
-<adding change notes with your prs>`, particularly the list of categories
+The pull request template to be consistent with :ref:`the contributing guidelines <adding change notes with your prs>`, particularly the list of categories
```
without this change the linter was failing.

webknjaz replied:
How?

csm10495 replied:
Same as CI: https://github.com/aio-libs/frozenlist/actions/runs/19160278857/job/54769109511

```
Checking dist/frozenlist-1.8.1.dev0-py3-none-any.whl: FAILED
ERROR    `long_description` has syntax errors in markup and would not be        
         rendered on PyPI.                                                      
         line 146: Warning: Inline literal start-string without end-string.     
Checking dist/frozenlist-1.8.1.dev0.tar.gz: FAILED
ERROR    `long_description` has syntax errors in markup and would not be        
         rendered on PyPI.                                                      
         line 146: Warning: Inline literal start-string without end-string.     
Error: Process completed with exit code 1.
```

webknjaz replied:
Oh, that must be my fault. Could you send this in a separate PR so I'd merge it right away?

csm10495 replied:
https://github.com/aio-libs/frozenlist/pull/719

webknjaz replied:
@csm10495 you can rebase the PR now this will disappear.

csm10495 committed:
Try to fix coverage/types

webknjaz commented on CHANGES/718.bugfix.rst:
```diff
@@ -0,0 +1,2 @@
+Fix bug where FrozenList could not be pickled/unpickled.
```
This would be a bit better but ideally, it should explain the effect the end-users would see in an observable way.
```suggestion
Fixed a bug where FrozenList could not be pickled/unpickled
```

csm10495 replied:
I'll rewrite this to make more sense.

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -1,11 +1,17 @@
+__all__ = (
+    "FrozenList",
+    "PyFrozenList",
+    "_reconstruct_pyfrozenlist",
```
I don't believe a private function should be exposed as public.

```suggestion
```

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,7 +79,43 @@ def __hash__(self):
         else:
             raise RuntimeError("Cannot hash unfrozen list.")
 
+    def __deepcopy__(self, memo: dict[int, Any]):
```
Use `object` whenever possible, avoid `Any`.
```suggestion
    def __deepcopy__(self, memo: dict[int, object]):
```

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,7 +79,43 @@ def __hash__(self):
+        )
+
+
+def _reconstruct_pyfrozenlist(items: list[Any], frozen: bool) -> "PyFrozenList":
```
```suggestion
def _reconstruct_pyfrozenlist(items: list[object], frozen: bool) -> "PyFrozenList":
```

webknjaz reviewed (commented):
I think this needs to introduce snapshot testing — a mechanic allowing us to be sure that pickling/unpickling remains working across different versions of frozenlist. The idea is that they'd be stored in-repo and so pickles created now would keep being checked for years to come. Multidict has something like that but it's too DIY.

Could you try some of the proper plugins?
* `pytest-harvest`:  https://smarie.github.io/python-pytest-harvest
* `syrupy`: https://til.simonwillison.net/pytest/syrupy / https://syrupy-project.github.io/syrupy
* `pytest-insta`: https://github.com/vberlier/pytest-insta
* `inline-snapshot`:  https://15r10nk.github.io/inline-snapshot

`syrupy` and `inline-snapshot` look quite active. But since I haven't played with them, I can't tell which one would feel more organic to use. Hence the need for some research.

webknjaz commented on tests/test_frozenlist.py:
```diff
@@ -372,6 +387,23 @@ def test_deepcopy_multiple_references(self) -> None:
         assert len(copied[1]) == 3  # Should see the change
         assert len(shared) == 2  # Original unchanged
 
+    @pytest.mark.parametrize("freeze", [True, False])
```
Could you also add human-readable IDs for context? This will make the test report easier to follow.

csm10495 commented:
> I think this needs to introduce snapshot testing — a mechanic allowing us to be sure that pickling/unpickling remains working across different versions of frozenlist. The idea is that they'd be stored in-repo and so pickles created now would keep being checked for years to come. Multidict has something like that but it's too DIY.
> 
> Could you try some of the proper plugins?
> 
> * `pytest-harvest`:  https://smarie.github.io/python-pytest-harvest
> * `syrupy`: https://til.simonwillison.net/pytest/syrupy / https://syrupy-project.github.io/syrupy
> * `pytest-insta`: https://github.com/vberlier/pytest-insta
> * `inline-snapshot`:  https://15r10nk.github.io/inline-snapshot
> 
> `syrupy` and `inline-snapshot` look quite active. But since I haven't played with them, I can't tell which one would feel more organic to use. Hence the need for some research.

I'm not fully sure I follow here. Pickling is safe when the serialized data is sent to newer python versions, but not necessarily older ones. I don't think we have a guarantee that we can pickle to binary then unpickle years from now if the structure changes. What types of guarantees are you looking to test?

csm10495 committed:
Update frozenlist/__init__.py

Co-authored-by: 🇺🇦 Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

csm10495 committed:
Update frozenlist/__init__.py

Co-authored-by: 🇺🇦 Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

csm10495 committed:
PR feedback

csm10495 committed:
PR feedback: add human readble ids

webknjaz commented:
> What types of guarantees are you looking to test?

Mostly that a pickle created with older frozenlist can be unpickled into the same structure with a newer one. Do you think it's an overkill for now?

webknjaz commented:
Looks like GH autoclosed this because of the keyword 😂. Reopening.

webknjaz commented on CHANGES/718.bugfix.rst:
```diff
@@ -0,0 +1,5 @@
+Fixed a bug where FrozenList could not be pickled/unpickled.
+
+Pickling is a requirement for FrozenList to be able to be passed to multiprocessing Processes. Without this
+fix, users receive a TypeError upon attempting to pickle a FrozenList.
```
```suggestion
fix, users receive a :exc:`TypeError` upon attempting to pickle a :class:`~frozenlist.FrozenList`.
```

webknjaz commented on CHANGES/718.bugfix.rst:
```diff
@@ -0,0 +1,5 @@
+Fixed a bug where FrozenList could not be pickled/unpickled.
+
+Pickling is a requirement for FrozenList to be able to be passed to multiprocessing Processes. Without this
```
```suggestion
Pickling is a requirement for :class:`~frozenlist.FrozenList` to be able to be passed to :class:`multiprocessing.Process`\ es. Without this
```

webknjaz commented on CHANGES/718.bugfix.rst:
```diff
@@ -0,0 +1,5 @@
+Fixed a bug where FrozenList could not be pickled/unpickled.
```
```suggestion
Fixed a bug where :class:`~frozenlist.FrozenList` could not be pickled/unpickled.
```

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -1,11 +1,15 @@
+__all__ = (
+    "FrozenList",
+    "PyFrozenList",
+)  # type: Tuple[str, ...]
```
Let's keep this unchanged.

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,7 +77,43 @@ def __hash__(self):
+        )
+
+
+def _reconstruct_pyfrozenlist(items: list[object], frozen: bool) -> "PyFrozenList":
```
If we put the definition before this function, there would be no need to quote the annotation. Let's make it so.
```suggestion
def _reconstruct_pyfrozenlist(items: list[object], frozen: bool) -> PyFrozenList:
```

csm10495 commented:
> > What types of guarantees are you looking to test?
> 
> Mostly that a pickle created with older frozenlist can be unpickled into the same structure with a newer one. Do you think it's an overkill for now?

I think its overkill since as long as the physical structure of frozenlist doesn't change in a non compatible way (like internal property names changing) it would be forwards compatible via pickle. In terms of other cases: I don't think libraries typically guarantee pickling cross versions anyways. 

The big use case is for serialization by multiprocessing.

csm10495 committed:
Apply suggestions from code review

Co-authored-by: 🇺🇦 Sviatoslav Sydorenko (Святослав Сидоренко) <wk.cvs.github@sydorenko.org.ua>

csm10495 committed:
More PR fixes

Dreamsorcerer commented on tests/test_frozenlist.py:
```diff
@@ -273,6 +274,20 @@ def test_deepcopy_frozen(self) -> None:
+        assert copied[2][0] == 1
+        assert copied[2][1] == 2
+        assert len(copied[2][2]) == 3
+        # ... and so on. Testing equality when a structure includes itself is tough.
```
Wouldn't `assert copied[2] is copied` accurately test this?

csm10495 replied:
Yes. I knew I was confusing myself. We deepcopy once, then its in the memo so it reuses at that and therefore references itself. Thanks!

Dreamsorcerer commented:
A forward compatibility test would be nice (though I won't insist). Such a test would look something like this:
https://github.com/aio-libs/aiohttp/blob/056d9297b2cb96c83466a3eaac3055acd9432229/tests/test_cookiejar.py#L1030

Dreamsorcerer commented on frozenlist/__init__.py:
```diff
@@ -73,10 +74,46 @@ def __hash__(self):
 
+        # Return existing copy if already processed (circular reference)
+        if obj_id in memo:
+            return memo[obj_id]
```
Codecov says this isn't covered. I think circular references are in the tests, so is this code actually needed? Or do we need a different test?

csm10495 replied:
Thanks for pointing this out. I was confused though think I figured it out. It looks like deepcopy itself handles the nested case via the memo before our code: https://github.com/python/cpython/blob/14c62227f9fa11fb743f9e03dcc5aab553de1098/Lib/copy.py#L125 .. Removed this bit.

csm10495 commented:
> A forward compatibility test would be nice (though I won't insist). Such a test would look something like this: https://github.com/aio-libs/aiohttp/blob/056d9297b2cb96c83466a3eaac3055acd9432229/tests/test_cookiejar.py#L1030

I'm not a huge fan of it but I'll add one. I'm don't think there are guarantees for binary format cross version normally.

csm10495 committed:
PR feedback. Add pickability forward compat test. Remove dead code for memo checking in deepcopy, use better way for self-including check test

Dreamsorcerer commented:
> I'm not a huge fan of it but I'll add one. I'm don't think there are guarantees for binary format cross version normally.

Yeah, unfortunately, that test comes from users complaining when we broke forward compatibility. We won't guarantee it, but we can atleast avoid it in patch releases and warn them in the changelog.

csm10495 committed:
General cleanup. Reverse the class order to allowing pickling to work a bit more natively.
Make the testing a bit smarter to auto skip CFrozenList tests if they don't apply.

webknjaz commented on CHANGES/718.bugfix.rst:
```diff
@@ -0,0 +1,5 @@
+
+Pickling is a requirement for :class:`~frozenlist.FrozenList` to be able to be passed to :class:`multiprocessing.Process`\ es. Without this
+fix, users receive a :exc:`TypeError` upon attempting to pickle a :class:`~frozenlist.FrozenList`.
+-- by :user:`csm10495`.
```
To make the byline detached, put it on a dedicated line and drop the trailing period.
```suggestion

-- by :user:`csm10495`
```

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,8 +74,36 @@ def __hash__(self):
+            {"_frozen": self._frozen},
+        )
+
+    def __setstate__(self, state: dict[str, object]):
```
```suggestion
    def __setstate__(self, state: dict[str, object]) -> None:
```

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,8 +74,36 @@ def __hash__(self):
+
+        return new_list
+
+    def __reduce__(self):
```
Add types for the return value?

webknjaz commented on frozenlist/__init__.py:
```diff
@@ -73,8 +74,36 @@ def __hash__(self):
         else:
             raise RuntimeError("Cannot hash unfrozen list.")
 
+    def __deepcopy__(self, memo: dict[int, object]):
```
```python
import typing as _t
```
```suggestion
    def __deepcopy__(self, memo: dict[int, object]) -> _t.Self:
```

csm10495 replied:
Considering this is a partial review, is it worth trying to fix things at this point?

Does it have a reasonable chance for merge?

webknjaz reviewed (commented):
I apparently had a few comments hanging as a draft for months. It's not a full review, though.
